### ISSUE TITLE
net::ERR_CONNECTION_REFUSED

### ISSUE THREAD

**Issue by falconbs** [tinyclues/saltpad#106]
I tried to install SaltPad and it seems to work until I get the Login screen. Pressing the login button nothing happens. In the developer console I see an error: 
"OPTIONS http://localhost:8000/login net::ERR_CONNECTION_REFUSED".

Could you give me a direction where to search for the solution? I used your proposed configuration except the port number which is mistyped. In the salt master config you wrote 8000 to the tornado port number, but in the settings.json you mentioned localhost:8050. This is the only difference.

Any idea?

**Comment by Lothiraldan** [comment]
Hi @falconbs, the salt-api is indeed configured to run on port 8000 but the VagrantFile is configured to redirect the guest port 8000 to host port 8050, I will choose a less used number and be coherent.

**Comment by falconbs** [comment]
Any idea why I get the above mentioned error?

**Comment by Lothiraldan** [comment]
Yes, as saltpad is a single-app page that runs into your browser, you should set in the configuration the port that will be available from your browser. Saltpad used to run on port 8000 inside the vagrant VM but was exposed on port 8050 on your host. Saltpad was configured to try to connect to salt-api on port 8050 because it's where is should be available.

I pushed a modification earlier so that salt-api should now listen on port 5417 on both the VM and your host to simplify the configuration.

**Comment by falconbs** [comment]
I have installed saltpad by hand as the readme instructs. Ubuntu 14.04 is used as server. I didn't used vagrant.
The error remained:
OPTIONS http://localhost:5417/login net::ERR_CONNECTION_REFUSED
e.exports @ xhr.js:134
(anonymous function) @ dispatchRequest.js:15
e.exports @ dispatchRequest.js:11

This is at the end of salt master config:
rest_tornado:
  port: 5417
  host: 0.0.0.0
  disable_ssl: true
  websockets: True
  cors_origin: '*'

This is nginx default site
server {
        listen 80 default_server;
        listen [::]:80 default_server ipv6only=on;
        root /opt/saltpad/;
        index index.html;
        server_name salt.falconsoft.hu;
        location / {
                try_files $uri /index.html;
        }
}

settings.json
{
    "API_URL": "localhost:5417",
    "SECURE_HTTP": false,
    "templates": {
        "basic": {
            "description": "Basic template",
            "matcher": "glob",
            "target": "*",
            "moduleFunction": "test.fib",
            "arguments": {
                "num": 10
            }
        },

        "version": {
            "description": "Basic version",
            "matcher": "glob",
            "target": "Celeste",
            "moduleFunction": "test.version",
            "arguments": {}
        }
    }
}

Testing with curl gives the following results:
root@salt:/etc/salt# curl -i -H accept=application/json -d username=saltadmin -d password=******* -d eauth=pam http://localhost:5417
HTTP/1.1 302 Found
Date: Fri, 08 Jan 2016 14:50:16 GMT
Content-Length: 0
Content-Type: text/html; charset=UTF-8
Location: /login
Server: TornadoServer/4.2.1

root@salt:/etc/salt# curl -i -H accept=application/json -d username=saltadmin -d password=******* -d eauth=pam http://192.168.168.186:5417
HTTP/1.1 302 Found
Date: Fri, 08 Jan 2016 14:50:27 GMT
Content-Length: 0
Content-Type: text/html; charset=UTF-8
Location: /login
Server: TornadoServer/4.2.1

**Comment by Lothiraldan** [comment]
If you don't use the vagrant environment, you need to point on your own development version of salt-api. Did you deployed one?

**Comment by falconbs** [comment]
I installed salt-api as it is in the readme.
"sudo apt-get install salt-api"
And configured as it is in the readme. (Edited salt master.)

What do you mean on "need to point on your own version"? To synchronize salt master "rest_tornado:" section with setting.json?

**Comment by falconbs** [comment]
Just an addition, running this commands work:
-----------------------------------------------------------------------
curl -sS localhost:5417/login \
    -H 'Accept: application/x-yaml' \
    -d username='saltadmin' \
    -d password='*******' \
    -d eauth='pam'

Result:
return:
- eauth: pam
  expire: 1452323065.611075
  perms: [.*, '@runner', '@wheel']
  start: 1452279865.611074
  token: 8e1418236f0e92b453373a8e9d21a05a
  user: saltadmin
-----------------------------------------------------------------------
curl -sS localhost:5417/run \
    -d client='local' \
    -d tgt='*' \
    -d fun='test.ping' \
    -d username='saltadmin' \
    -d password='*******' \
    -d eauth='pam' 

Result:
{"return": [{"master-minion": true}]}
-----------------------------------------------------------------------

**Comment by falconbs** [comment]
I installed a VirtualBox szerver with the help of vagrant. It works.
I compared every config file and saltpad itself.
1. salt master config is the same (except the used name).
2. nginx -conf and site are the same.
3. /opt/saltpad files are the same.
4. I checked the login through salt-api with curl. 

When I try to login the result status is 401 Unauthorized. What can be the difference between the two system?

**Comment by falconbs** [comment]
It works finally. The solution: I have to user salt-api 2015.8.0-1202-g0832539 (Beryllium) which is used in the vagrant test environment.
Originally I used  http://repo.saltstack.com/apt/ubuntu/14.04/amd64/latest repository to install Salt. It installed salt-api 2015.8.3 and it caused the problem.
I copied the /usr/lib/python2.7/dist-packages/salt directory over my installation because downgrade is not possible and now everything works.

Conclusion: SaltPad is not working with the latest salt-api 2015.8.3. I suppose there is a problem with the CORS config or CORS handling.

Anyways, thanks for your help.
